src/t_hash.c

```diff
@@ -1220,36 +1896,47 @@ void hashTypeRandomElement(robj *hashobj, unsigned long hashsize, listpackEntry
 static ExpireAction hashTypeActiveExpire(eItem _hashObj, void *ctx) {
     robj *hashObj = (robj *) _hashObj;
     ActiveExpireCtx *activeExpireCtx = (ActiveExpireCtx *) ctx;
+    sds keystr = NULL;
+    ExpireInfo info = {0};
 
     /* If no more quota left for this callback, stop */
     if (activeExpireCtx->fieldsToExpireQuota == 0)
         return ACT_STOP_ACTIVE_EXP;
 
-    if (hashObj->encoding == OBJ_ENCODING_LISTPACK) {
-        serverPanic("Listpack encoding not supported yet");
-    }
-    serverAssert(hashObj->encoding == OBJ_ENCODING_HT);
+    if (hashObj->encoding == OBJ_ENCODING_LISTPACK_EX) {
+        info = (ExpireInfo){
+                .maxToExpire = activeExpireCtx->fieldsToExpireQuota,
+                .ctx = hashObj,
+                .now = commandTimeSnapshot(),
+                .itemsExpired = 0};
 
-    dict *d = hashObj->ptr;
-    dictExpireMetadata *dictExpireMeta = (dictExpireMetadata *) dictMetadata(d);
+        listpackExExpire(hashObj, &info, &server.stat_expired_hash_fields);
```

Comment:
this allows listpackExExpire() to be more independent.
```suggestion
        listpackExExpire(hashObj, &info);
        server.stat_expired_hash_fields += info.itemsExpired;
```

Reply:
I agree with the comment, but just to verify: `listpackExExpire` was increasing `stat_expired_hash_fields` internally before this change. So after my fix it will not increase ANY stat value internally, instead its callers will increase whichever stat value is right after calling `listpackExExpire`, based on the value returned in `info.itemsExpired`. Is this what you mean?

Reply:
yes, this decouples listpackExExpire() from the statistics.